lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -3653,17 +3654,19 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     return (insertion.first->getSecond() = bufWithCleanup);
   }
 
-  void addToAdjointBuffer(SILValue originalBuffer,
+  void addToAdjointBuffer(SILValue destBuffer,
```

Comment:
I wonder if there's a better name than `addToAdjointBuffer`?
The name is not ideal because `destBuffer` may not in fact be an "adjoint buffer" (i.e. something stored in `AdjointEmitter::bufferMap`).

A more accurate name would be like "add to buffer in adjoint function". Another idea is to remove `addToAdjointBuffer` now that it simply forwards `accumulateIndirect` - but I would like to keep the function if it can be implemented with correct semantics, because it reduces caller boilerplate.

@rxwei: what are your thoughts?

Reply:
Before talking about names, `addToAdjointBuffer` is supposed to take an original buffer and accumulate the new value into the adjoint buffer corresponding to it. Why did you change the behavior here?

Reply:
Ah, that's another way to look at things. I agree with your view now.
How about this:
* Change `addToAdjointBuffer` back to the old behavior, but add assertion that `originalBuffer->getFunction() == &getAdjoint()`.
* Change current call to `addToAdjointBuffer` in `visitStructElementAddrInst` to use `accumulateIndirect`.

Reply:
> * Change current call to `addToAdjointBuffer` in `visitStructElementAddrInst` to use `accumulateIndirect`.

Can you elaborate on whether `addToAdjointBuffer` doesn't work? You should call `addToAdjointBuffer` on the original operand of `visitStructElementAddrInst`.

Reply:
> Can you elaborate on why `addToAdjointBuffer` doesn't work? You should call `addToAdjointBuffer` on the original operand of `visitStructElementAddrInst`.

Good question. The answer lies in the current transformation rule for `struct_element_addr`:
```
// Original instruction.
y = struct_element_addr x, #key
// Adjoint transformation.
adj[x]#key' = struct_element_addr adj[x], #key'
adj[x]#key' += adj[y]
// Note that `adj[x] += ...` isn't called.
```

Does this make sense to you?

Reply:
We discussed in person and settled on this design (mentioned above):
> * Change `addToAdjointBuffer` back to the old behavior, but add assertion that `originalBuffer->getFunction() == &getAdjoint()`.
> * Change current call to `addToAdjointBuffer` in `visitStructElementAddrInst` to use `accumulateIndirect`.

This is done in https://github.com/apple/swift/pull/22669/commits/65b8c28603d2e5a31dfd62186a90a26f34639e40.